katas/content/solving_graph_coloring/weak_coloring_one_vertex/solution.md

```diff
@@ -0,0 +1,12 @@
+To implement this check, you need to start by counting the vertices connected to the given vertex 
+and allocating an array of auxiliary qubits, one for each neighboring vertex. These qubits will have their states flipped if the vertices connected by the corresponding edge have the same color. 
+This can be done easily using with the operation `Oracle_ColorEquality` from an earlier task, similarly to how you did in when validating the vertex coloring of the graph. 
+
+Now, for the coloring of the vertex to be valid, it needs to either have no neighbors, or to have at least one neighbor of a different color, so the array of auxiliary qubits should be in any state but $\ket{1...1}$. To implement this, you can flip the target qubit $\ket{y}$ unconditionally using the $X$ gate, and then flip it again only if qubits are in state $\ket{1...1}$.
```

Comment:
The references to state $\ket{1...1}$ are confusing to me here because the solution uses `ApplyControlledOnInt(0, X, x1, y);` which I take to mean that it will flip it again only when the qubits are in state $\ket{0...0}$. Am I misunderstanding this?

Reply:
The part of the solution this paragraph talks about is in `Oracle_WeakColoring_OneVertex`:
```
            X(y);
            if nNeighbors > 0 {
                Controlled X(sameColorChecks, y);
            }
```
`ApplyControlledOnInt(0)` shows up in `Oracle_ColorEquality` that is subroutine from an earlier exercise, so I don't take about it here

Reply:
Yes, as a subroutine of the solution, it seems like the solution is only flipping the output when the input qubits are in state $\ket{0...0}$.

Reply:
Sorry, I'm not sure I follow...
When we compare two colors, we use a different logic indeed: since we're computing XORs of pairs of bits to check that they're all 0, we get the results by checking that all qubits are 0. But this was discussed earlier, in exercise "Are Colors Equal?", and I'm assuming that the reader is already familiar with this

Reply:
Oh if flips the solution's output when all of the individual color checks return $\ket{1}$. Is that right?

Reply:
Yes, with a twist: if all of the individual color checks return $\ket{1}$, that's the only scenario when weak coloring is *in*valid (meaning all neighbors' colors are equal; at least one of color pairs has to be different and return $\ket{0}$ for it to be valid). 

So the complete solution flips the output twice: once unconditionally, and once if all of the individual color checks return $\ket{1}$. This means that all the individual color checks return $\ket{1}$ is the only scenario in which the result remains unflipped, indicating invalid coloring

Reply:
That makes sense. Thanks!